src/index.ts

```diff
@@ -1458,4 +1502,111 @@ export function VerifyWitness(
     }
   }
   throw new VerificationError(`Unexpected end of proof`);
-}
\ No newline at end of file
+}
+
+/**
+ * Verifies a witness against a staleRoot, constructs newest witness from the recentState and verifies it
+ * @param staleRoot   stale root
+ * @param key         key being read
+ * @param witness     witness against a stale root
+ * @param recentState Merkle Paticia Tree horizontally caches nodes to some depth and veritically caches most recent keys
+ */
+export function verifyStaleWitness(
+    staleRoot: Buffer, key: Buffer, witness: RlpWitness, recentState: MerklePatriciaTree) {
+  // Verify if the witness is valid against the stale root
+  verifyWitness(staleRoot, key, witness);
+  // If the stale root and recent root match, return the witness
+  if (staleRoot.compare(recentState.root) === 0) {
+    return witness;
+  }
+  // Check if nodes match at any feasible depth (search happens till available maxCacheDepth)
+  const recentWitness: Witness<Buffer> = {proof: [], value: null};
+  const result: SearchResult = recentState.search(key);
+  // If the key was vertically cached in the recentState; search will find the path along the key
+  if (result.remainder.length === 0 && result.node !== null) {
+    if (result.node.value!.compare(witness.value!) !== 0) {
+      throw new Error("key has been updated, witness cannot be used");
+    }
+    return;
+  }
+  // If it was not vertically cached, then the key wasn't changed so the node hashes should match at some depth
+  recentWitness.proof = [];
+  recentWitness.value = witness.value;
+  const oldNodeHashes: Array<bigint> = [];
+  for (const serializedOldNode of witness.proof) {
+    oldNodeHashes.push(hashAsBigInt(HashType.KECCAK256, serializedOldNode));
+  }
+  for (const [idx, witNode] of result.stack.entries()) {
+    let recentHash: bigint;
+    recentWitness.proof.push(witNode);
+    recentHash = witNode.hash(options as {} as MerklePatriciaTreeOptions<{}, Buffer>);
+    for (let j = 0; j < oldNodeHashes.length; j++) {
+      if (recentHash === oldNodeHashes[j]) {
+        const curWit = recentState.rlpSerializeWitness(recentWitness);
+        for(j = j+1; j < oldNodeHashes.length; j++) {
+          curWit.proof.push(witness.proof[j]);
+        }
+        verifyWitness(recentState.root, key, curWit);
+        return;
+      }
+    }
+  }
+  throw new VerificationError("stale witness verification failed");
+}
+
+export class CachedMerklePatriciaTree extends MerklePatriciaTree<Buffer, Buffer> {
+  // Maximum depth of the cached MerklePatriciaTree with rootNode is at a depth 1.
+  private maxCacheDepth: number;
+
+  constructor (depth = 6) {
+    super();
+    // Set the default maxCacheDepth to 6
+    this.maxCacheDepth = depth;
+  }
+
+  /** Returns the maxCacheDepth */
+  getmaxCacheDepth () : number {
+    return this.maxCacheDepth;
+  }
+
```

Comment:
Where's the putWitness function we discussed? We want to take a Witness, verifying it against the cached tree. You can assume that the witness is RLP encoded.

Reply:
Yeah I'm working on it, trying to test and fix it. Will be done in an hour or so.

Reply:
I think putWitness shouldn't perform any verification, but just populate the cachedState right? And then we verify the stale witness using this populated cachedState.
We cannot really figure out if we want to add a witness to the cache or not just looking at the witness and the root right? We don't know if root is stale or more recent.

Reply:
It should perform verification, otherwise it will add potentially incorrect data!